compiler/generator_rust.c

```diff
@@ -0,0 +1,1355 @@
+        {
+            g->I[0] = i;
+            g->I[1] = v->i;
+            g->I[2] = v->result;
+            g->L[0] = v->b;
+            g->S[0] = i < x->literalstring_count - 1 ? "," : "";
+
+            w(g, "~MAmong(~L0, ~I1, ~I2, None");
+            if (v->function != 0)
+            {
+                w(g, ", \"");
+                if (v->function->type == t_routine) {
+                    /* Need to use mangled version of private name here. */
+                    w(g, "_~n");
+                }
+                write_varname(g, v->function);
+                w(g, "\"");
+            }
+            w(g, ")~S0~N");
+            v++;
+        }
+    }
+    w(g, "~-~M];~N~N");
+}
+
+static void generate_amongs(struct generator * g) {
+
+    struct among * x = g->analyser->amongs;
+    while (x != 0) {
+        generate_among_table(g, x);
+        x = x->next;
+    }
+}
+
+static void set_bit(symbol * b, int i) { b[i/8] |= 1 << i%8; }
+
+static void generate_grouping_table(struct generator * g, struct grouping * q) {
+
+    int range = q->largest_ch - q->smallest_ch + 1;
+    int size = (range + 7)/ 8;  /* assume 8 bits per symbol */
+    symbol * b = q->b;
+    symbol * map = create_b(size);
+    int i;
+    for (i = 0; i < size; i++) map[i] = 0;
+
+    /* Using unicode would require revision here */
+
+    for (i = 0; i < SIZE(b); i++) set_bit(map, b[i] - q->smallest_ch);
+
+    g->V[0] = q->name;
+    g->I[0] = size;
+    w(g, "~Mstatic ~W0: &'static [u8; ~I0] = &[");
+    for (i = 0; i < size; i++) {
+        write_int(g, map[i]);
+        if (i < size - 1) w(g, ", ");
+    }
+    w(g, "];~N~N");
+    lose_b(map);
+}
+
+static void generate_groupings(struct generator * g) {
+    struct grouping * q;
+    for (q = g->analyser->groupings; q; q = q->next) {
+        generate_grouping_table(g, q);
+    }
+}
+
+
+static void generate_members(struct generator * g) {
+  
+    struct name * q;
+    w(g, "struct Context {~+~N");
+    for (q = g->analyser->names; q; q = q->next) {
+        g->V[0] = q;
+        switch (q->type) {
+            case t_string:
+                w(g, "~M~W0: &'static str,~N");
+                break;
+            case t_integer:
+                w(g, "~M~W0: usize,~N");
+                break;
+            case t_boolean:
+                w(g, "~M~W0: bool,~N");
+                break;
+        }
+    }
+    w(g, "~-}~N");
+}
+
+static void generate_copyfrom(struct generator * g) {
+
+    struct name * q;
+    if (g->copy_from_count == 0) return;
+    w(g, "~Mdef copy_from(self, other):~+~N");
+    for (q = g->analyser->names; q != 0; q = q->next) {
+        g->V[0] = q;
+        switch (q->type) {
+            case t_string:
+            case t_integer:
+            case t_boolean:
+                w(g, "~M~V0 = other.~W0~N");
+                break;
+        }
+    }
+    /* For Python 3, this can just be super(). */
+    w(g, "~Msuper(~n, self).copy_from(other)~N~-");
```

Comment:
I solved this now by both cloning `env` and `context` in `generate_dollar`. 
My testcase compiled. Don't know if it behaves as intended, though.
This is now similar to the c solution AFAIK.